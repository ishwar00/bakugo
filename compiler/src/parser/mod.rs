use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
