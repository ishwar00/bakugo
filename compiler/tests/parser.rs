use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
