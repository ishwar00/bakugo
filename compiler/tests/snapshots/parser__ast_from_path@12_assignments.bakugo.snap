---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/12_assignments.bakugo
---
SourceFile {
    top_level: [
        FnDecl(
            FnDecl {
                name: Ident {
                    value: "main",
                    span: Span {
                        str: "main",
                        start: 5,
                        end: 9,
                    },
                },
                params: [],
                result_kind: None,
                body: StatementList(
                    [
                        Assignment {
                            left: Identifier {
                                value: "a",
                                span: Span {
                                    str: "a",
                                    start: 15,
                                    end: 16,
                                },
                            },
                            right: Integer {
                                value: 45,
                                span: Span {
                                    str: "45",
                                    start: 19,
                                    end: 21,
                                },
                            },
                        },
                        Assignment {
                            left: Identifier {
                                value: "a",
                                span: Span {
                                    str: "a",
                                    start: 24,
                                    end: 25,
                                },
                            },
                            right: Integer {
                                value: 45,
                                span: Span {
                                    str: "45",
                                    start: 34,
                                    end: 36,
                                },
                            },
                        },
                        Assignment {
                            left: Identifier {
                                value: "b",
                                span: Span {
                                    str: "b",
                                    start: 27,
                                    end: 28,
                                },
                            },
                            right: Integer {
                                value: 2,
                                span: Span {
                                    str: "2",
                                    start: 38,
                                    end: 39,
                                },
                            },
                        },
                        Assignment {
                            left: Identifier {
                                value: "c",
                                span: Span {
                                    str: "c",
                                    start: 30,
                                    end: 31,
                                },
                            },
                            right: Integer {
                                value: 3,
                                span: Span {
                                    str: "3",
                                    start: 41,
                                    end: 42,
                                },
                            },
                        },
                    ],
                ),
                span: Span {
                    str: "func main() {\n\ta = 45;\n\ta, b, c = 45, 2, 3;\n}",
                    start: 0,
                    end: 45,
                },
            },
        ),
    ],
}
